File: backend/app/services/official_api_base_routes.py
from __future__ import annotations

import json
import re
from functools import lru_cache
from pathlib import Path
from typing import Any

_SAFE_REGION = re.compile(r"^[A-Za-z0-9][A-Za-z0-9.-]{0,79}$")


@lru_cache(maxsize=1)
def _catalog() -> dict[str, Any]:
    path = Path(__file__).resolve().parents[3] / "policies" / "official-api-base-routes.json"
    payload = json.loads(path.read_text(encoding="utf-8"))
    if payload.get("schema_version") != "astraquote-official-api-base-routes/1":
        raise RuntimeError("AstraQuote official API base-route catalog is invalid")
    return payload


def _service_key(value: str) -> str:
    return re.sub(r"[^a-z0-9]+", "", str(value).strip().casefold())


def official_api_base_route(
    provider: str, service: str, region: str
) -> dict[str, str] | None:
    """Return only a verified official hostname for a provider service.

    Paths, actions, parameters, response shapes and rates deliberately remain
    live-query inputs.  This prevents the old detailed-route cache from growing
    back while still stopping callers from inventing provider hostnames.
    """

    provider_key = str(provider).strip().casefold()
    service_key = _service_key(service)
    region_value = str(region).strip()
    if not service_key or not _SAFE_REGION.fullmatch(region_value):
        return None
    provider_routes = _catalog().get("providers", {}).get(provider_key)
    if not isinstance(provider_routes, dict):
        return None
    selected: dict[str, Any] | None = None
    for candidate in provider_routes.get("routes") or []:
        if not isinstance(candidate, dict):
            continue
        aliases = {_service_key(alias) for alias in candidate.get("services") or []}
        if service_key in aliases:
            selected = candidate
            break
    if selected is None and provider_routes.get("default_endpoint"):
        selected = provider_routes
    if selected is None:
        return None
    endpoint = str(
        selected.get("endpoint") or selected.get("default_endpoint") or ""
    ).strip().casefold()
    endpoint_template = str(selected.get("endpoint_template") or "").strip().casefold()
    if not endpoint and endpoint_template:
        endpoint = endpoint_template.replace("{region}", region_value.casefold())
    if not endpoint:
        return None
    source_url = str(
        selected.get("official_source_url")
        or provider_routes.get("official_source_url")
        or ""
    ).strip()
    return {
        "provider": provider_key,
        "service": service_key,
        "endpoint": endpoint,
        "official_source_url": source_url,
        "catalog_checked_at": str(_catalog().get("catalog_checked_at") or ""),
    }

File: deploy/astraquote-mcp/lib/official-api-base-routes.js
'use strict';

const fs = require('node:fs');
const path = require('node:path');


const CATALOG_CANDIDATES = [
  process.env.ASTRAQUOTE_OFFICIAL_API_BASE_ROUTES_PATH,
  path.resolve(__dirname, '../policies/official-api-base-routes.json'),
  path.resolve(__dirname, '../../policies/official-api-base-routes.json'),
  path.resolve(__dirname, '../../../policies/official-api-base-routes.json'),
].filter(Boolean);
const SAFE_REGION = /^[A-Za-z0-9][A-Za-z0-9.-]{0,79}$/;
const MUTATING_OPERATION_VERBS = new Set([
  'activate', 'add', 'allocate', 'apply', 'attach', 'bind', 'cancel', 'change',
  'close', 'create', 'deactivate', 'delete', 'deploy', 'disable', 'downgrade',
  'detach', 'enable', 'execute', 'install', 'invoke', 'launch', 'modify', 'open',
  'pay', 'purchase', 'buy', 'reboot',
  'refund', 'release', 'remove', 'renew', 'reset', 'resize', 'restart', 'run',
  'scale', 'set', 'start', 'stop', 'submit', 'subscribe', 'terminate', 'unbind',
  'uninstall', 'unsubscribe', 'update', 'upgrade',
]);

let cachedCatalog;


function catalog() {
  if (cachedCatalog !== undefined) return cachedCatalog;
  for (const candidatePath of CATALOG_CANDIDATES) {
    try {
      const payload = JSON.parse(fs.readFileSync(candidatePath, 'utf8'));
      if (payload?.schema_version === 'astraquote-official-api-base-routes/2') {
        cachedCatalog = payload;
        return payload;
      }
    } catch {
      // Try the next supported source/deployment layout.
    }
  }
  cachedCatalog = null;
  return null;
}


function serviceKey(value) {
  return String(value || '').trim().toLowerCase().replace(/[^a-z0-9]+/g, '');
}


function operationIdentity(query) {
  if (String(query?.action || '').trim()) return String(query.action).trim();
  const segments = String(query?.path || '').split('/').filter(Boolean);
  return segments.at(-1) || '';
}


function operationVerb(value) {
  const tokens = String(value || '')
    .replace(/([a-z0-9])([A-Z])/g, '$1 $2')
    .split(/[^A-Za-z0-9]+/u)
    .map((token) => token.toLowerCase())
    .filter((token) => /^[a-z]/u.test(token));
  const first = tokens[0] || '';
  if (first === 'batch') return tokens[1] || first;
  if (first.startsWith('batch') && first.length > 5) return first.slice(5);
  return first;
}


function assertNonMutatingOperation(query, route, registered) {
  const operation = operationIdentity(query);
  if (!operation) {
    const error = new Error('An official API operation is required.');
    error.code = 'official_api_operation_missing';
    error.retryable = true;
    error.details = {
      provider: route.provider,
      service: route.service,
      registered_operations: route.operations,
    };
    throw error;
  }
  // Catalog operations are trusted defaults, not a closed allowlist. Providers
  // add read-only discovery and pricing actions over time, so pass those through
  // while retaining the official-host boundary and blocking state changes.
  if (registered || !MUTATING_OPERATION_VERBS.has(operationVerb(operation))) return;
  const error = new Error('State-changing official API operations are blocked.');
  error.code = 'official_api_mutating_operation_blocked';
  error.retryable = false;
  error.details = {
    provider: route.provider,
    service: route.service,
    blocked_operation: operation,
  };
  throw error;
}


function officialApiBaseRoute(provider, service, region) {
  const providerKey = String(provider || '').trim().toLowerCase();
  const normalizedService = serviceKey(service);
  const regionValue = String(region || '').trim();
  if (!normalizedService || !SAFE_REGION.test(regionValue)) return null;
  const providerRoutes = catalog()?.providers?.[providerKey];
  if (!providerRoutes || typeof providerRoutes !== 'object') return null;
  const selected = (providerRoutes.routes || []).find((candidate) => (
    (candidate.services || []).some((alias) => serviceKey(alias) === normalizedService)
  )) || (
    providerRoutes.default_route && typeof providerRoutes.default_route === 'object'
      ? providerRoutes.default_route
      : null
  );
  if (!selected) return null;
  const capability = String(selected.capability || 'quote_api').trim();
  if (!['quote_api', 'official_page_only'].includes(capability)) return null;
  const endpoint = String(selected.endpoint || selected.endpoint_template || '')
    .trim().toLowerCase().replace('{region}', regionValue.toLowerCase());
  if (capability === 'quote_api' && !endpoint) return null;
  return {
    provider: providerKey,
    service: normalizedService,
    endpoint,
    capability,
    operations: (selected.operations || [])
      .map((operation) => String(operation).trim())
      .filter(Boolean),
    official_source_url: String(
      selected.official_source_url || providerRoutes.official_source_url || '',
    ).trim(),
    catalog_checked_at: String(catalog()?.catalog_checked_at || ''),
  };
}


function withOfficialApiBaseRoute(query) {
  const route = officialApiBaseRoute(query?.provider, query?.service, query?.region);
  if (!route) return { ...query };
  const materialized = { ...query };
  if (route.capability === 'quote_api') materialized.endpoint = route.endpoint;
  else delete materialized.endpoint;
  if (!materialized.official_source_url && route.official_source_url) {
    materialized.official_source_url = route.official_source_url;
  }
  if (route.capability === 'quote_api' && route.operations.length > 0) {
    if (!materialized.action && (!materialized.path || materialized.path === '/')
      && route.operations.length === 1) {
      const operation = route.operations[0];
      if (operation.startsWith('/')) materialized.path = operation;
      else materialized.action = operation;
    }
    const supplied = String(materialized.action || materialized.path || '/').toLowerCase()
      .replace(/\/$/u, '');
    const allowed = route.operations.some((operation) => {
      const expected = operation.toLowerCase().replace(/\/$/u, '');
      return supplied === expected
        || (!materialized.action && !operation.startsWith('/') && supplied.endsWith(expected));
    });
    assertNonMutatingOperation(materialized, route, allowed);
  }
  return {
    ...materialized,
  };
}


module.exports = { officialApiBaseRoute, withOfficialApiBaseRoute };
